Revise librarycontrol // FocusWidget

ronso0 opened the pull request:
While testing #4465 I was looking for a way to reliably navigate the track menu with a push encoder, and it turned out my `FocusWidget` approach in #4369 is complicated and not extendable to cover menus and dialogs.

Here I revise the focus control:
* hook up to QApplication's `focusChanged` and `focusWindowChanged` signals. These are emitted every time the focused widget changes or when a new window/popup/menu is ready to receive keyboard presses, thus I could remove almost the entire former implementation.
* these signals both invoke `updateFocusedWidgetControls()`
* this identifies the currently focused Mixxx widget, QDialog or QMenu, and stores the `FocusWidget` type, as well as the respective new state of `[Library],focused_widget`. 
* this `FocusWidget` is now used by a few library controls to trigger context-specific actions -- this and `LibraryControl::emitKeyEvent` is now much simpler, faster and less error-prone I reckon

Updated enum values of `[Library],focused_widget`:
``` c++
0  None        r
1  Searchbar   rw
2  Sidebar     rw
3  TracksTable rw  // or a feature root view (WLibraryTextBrowser)
4  ContextMenu r   // any context menu + main menu bar
                   // effect selector popup + AutoDJ transition mode popup
5  Dialog      r   // any file or error dialog, CoverArt, key wheel window, preferences, About
6  Unknown     r   // Unknown skin widget or unknown window
```

### TESTING
with controller:
Every `[Library]` control should still work as expected.
For testing the menu navigation, manually open a context menu or one in the main menu bar, then use `[Library],MoveVertical` to move the selection, `[Library],GoToItem` to activate items and confirm dialogs. Or put https://github.com/ronso0/mixxx/tree/open-track-menu-control on top of this branch.
It's also possible to handle error messages and some compact dialogs with these controls, too.

without controller:
open developer tools, find `focused_widget`, then click around in the main GUI and watch how the value changes accordingly, see enum above.

ronso0 committed:
librarycontrol: consolidate move.. and scroll.. functions

ronso0 committed:
librarycontrol: revise focus control

ronso0 committed:
librarycontrol: use FocusWidget for MoveVertical & GoToItem, revise emitKeyEvent

ronso0 committed:
Library: rename loadSelectedTrack > activateSelectedTrack

ronso0 committed:
librarycontrol: simplify slotSelectSidebarItem

ronso0 commented:
Alright, I tested all controls that were touched by this PR, as well as navigating menus and dialogs.
Everything works as expected :tada:

daschuer commented on src/library/librarycontrol.cpp:
```diff
@@ -324,7 +325,23 @@ LibraryControl::LibraryControl(Library* pLibrary)
+    // Show the track context menu for selected tracks, or hide it
+    // if it is the current active window
+    m_pShowTrackMenu = std::make_unique<ControlPushButton>(
+            ConfigKey("[Library]", "show_track_menu"));
```
how about "toggle_track_menu" or "show_hide_track_menu"

ronso0 replied:
I removed that commit, it's now only in #4465

daschuer commented:
There seems to be a kind of loop when the Mixxx window receives the focus, resulting in a flashing focus border. 
Is this intended?

daschuer commented:
The rest works nice. Please give a hint when you consider this a ready for merge.

ronso0 commented:
> There seems to be a kind of loop when the Mixxx window receives the focus, resulting in a flashing focus border. Is this intended?

Not intended at all, and I can't reproduce.
I only hook up to the focus signals, not override focusEvents, so I don't know how the flashing is related to my changes (alternating focusIn/focusOut events). Mixxx itself does not set focus itself -- `setLibraryFocus(FocusWidget)` is only called when you press `Esc` in the sidebar or in the search box.

Can you describe your scenario and which widget is flashing?

Swiftb0y commented:
> There seems to be a kind of loop when the Mixxx window receives the focus, resulting in a flashing focus border. Is this intended?

fwiw, I can't reproduce this either.

daschuer commented:
Oh I have just notice that other Qt Apps are affected as well. So this it probably a Qt issue.

ronso0 commented:
Please let me know when someone starts a code review.
Until then I may still force-push with cosmetic changes / fixups not worth separate commits.

ronso0 committed:
librarycontrol: use 'BackTab' instead of 'Shift+Tab' ...

Emulated Shift+Tab sent to QApplication::focusWidget() moves the focus,
THOUGH it also makes Shift get stuck for this widget.
Actually, this wouldn't matter because Shift is released as soon as this
widget receives any keyEvent with Qt::NoModifier.
Though, [Playlist],selectTrackKnob doesn't use keyEvents, thus using it
directly after [Library],MoveFocusBackward would expand the selection...

Simply use BackTab key :)

Holzhaus commented on src/widget/wtracktableview.h:
```diff
@@ -35,7 +34,7 @@ class WTrackTableView : public WLibraryTableView {
     void setFocus() override;
     void keyPressEvent(QKeyEvent* event) override;
-    void loadSelectedTrack() override;
+    void doubleClickSelectedTrack() override;
```
We shouldn't name signals like that. Double clicking only works computers with a mouse. It does not work on touchscreens (double tapping), it does not work on controllers (long press?).

How about something like `triggerSelectedTrackSecondaryAction()` or something like that?

Holzhaus replied:
Or is this a non-issue because this signal will never be emitted by using a touchscreen or a controller?

ronso0 replied:
I simply adjusted to the wording in Preferences > Library > **Track Double-Click Action** so that the connection is clear for developers.

* Controller mappings would use [`[Library], GoToItem`](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-[Library]-GoToItem)
* Keyboards: #4359
* I don't remember testing with a touch screen tbh, no idea how one can interact with the library. Depends on the OS and desktop environment IIRC.

ywwg replied:
suggestion:  activateSelectedTrack

ronso0 replied:
yup, perfect!

ronso0 committed:
Merge remote-tracking branch 'mixxx/main' into focus-widget-library-control

ronso0 commented:
For testing without controllers you can put abd1520830 on top to have some debug output and skin feedback for FocusWidget.

daschuer reviewed (approved):
LGTM, And works as expected. Thank you.

ronso0 commented:
Great, I'll add the documentation when I'm back home.

ronso0 commented:
fixup https://github.com/mixxxdj/mixxx/commit/4f45dbe2ea8227263060b8919665c19f3bb7374c
Thanks @JosepMaJAZ for spotting
